### PR TITLE
feat(journal): removal of intermediate serialisation in raft append

### LINKED ISSUE #6671: Remove unnecessary intermediate serialization in RaftLog#append
**Description**

At the moment, we first serialize the Raft log entry before passing it to the journal. This works fine, but requires us to have an intermediate in memory buffer. This means we have an extra copy in the journal when appending, and we have this expendable, long living buffer in memory.

Proposal:

What we need is a way to serialize only when we have the final destination buffer. One easy way is to amend the Journal interface to add a `public JournalRecord append(int entryLength, Consumer<MutableDirectBuffer> entryWriter);`. Or we we want to obviate the need for lambdas:

```java
public interface RecordDataWriter<T> {
  final void write(T entry, MutableDirectBuffer buffer, int offset);
}

public interface Journal {
  // ...
  public <T> JournalRecord append(T entry, int entryLength, RecordDataWriter<T> writer)
}
```

In the journal, we can then serialize the record data when we need it directly to the writer's buffer instead of using an intermediate write buffer.

NOTE: if we can make it without using generics, even better :slightly_smiling_face: 

/cc @deepthidevaki

### PR DESCRIPTION
**Pull request by megglos**
## Description

This change removes the intermediate buffer previously used in `RaftLog` to serialise the raft log entry before passing it to the journal where it is written to the journal log. Instead a recordWriter instance is passed to the journal which holds a reference to the entry and directly writes the data to the writeBuffer when the actual journal log entry is written.

## Related issues

closes #6671 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark
http://34.77.165.228/d/I4lo7_EZk/zeebe?orgId=1&refresh=10s&var-DS_PROMETHEUS=Prometheus&var-namespace=meg-6671&var-pod=All&var-partition=All

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: journal/src/main/java/io/camunda/zeebe/journal/Journal.java [hunk 1]
```diff
@@ -17,30 +17,30 @@
 
 import io.camunda.zeebe.journal.JournalException.InvalidChecksum;
 import io.camunda.zeebe.journal.JournalException.InvalidIndex;
-import org.agrona.DirectBuffer;
 
 public interface Journal extends AutoCloseable {
```
Comment (megglos) [inline journal/src/main/java/io/camunda/zeebe/journal/Journal.java:21]: I went for reducing the number of overloaded append methods here as all the ones that accepted DirectBuffers were used by tests exclusively. This comes at the cost of a bit more boilerplate in the tests but actually makes sure the exact methods used by production code are tested.

File: journal/src/main/java/io/camunda/zeebe/journal/RecordDataWriter.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.journal;
+
+import org.agrona.MutableDirectBuffer;
+
+public interface RecordDataWriter {
```
Comment (npepinpe) [inline journal/src/main/java/io/camunda/zeebe/journal/RecordDataWriter.java:12]: :x: I believe this is a duplicate of `BufferWriter`, so we could simply use this, unless there is a semantic reason to differentiate between either?

**Review by npepinpe — COMMENTED** [review]
:+1: 

If I understand correctly, the idea is to not pass a buffer, but to pass a serialization strategy instead? Something like a type which knows how to self-serialize?

I've never been a big fan of this approach (I'm aware we use it heavily in Zeebe :smile:), and would rather split serialization strategy from data, but I can see how here it would become somewhat cumbersome (you'd have to pass the instance and it's serializer around all the way down to the final write operation). I think it's worth exploring in a time-boxed fashion, because maybe I'm overestimating the ugliness of the approach...perhaps you've already done that, I'd be curious to hear what came out of it. Anyway, ignoring this as I agree it's probably not ideal with our current design...

We could add a method in `RaftEntry` (which all concrete entries implement), e.g. `BufferWriter toSerializable(int,RaftEntrySerializer)`, which could return a simple adapter encapsulating the `RaftEntrySerializer` and the entry itself. This avoids having to iterate of the possible types, and delegates to the type itself to produce the adapter.

e.g. `ApplicationEntry`

```java
public record ApplicationEntry(long lowestPosition, long highestPosition, DirectBuffer data)
    implements RaftEntry {

  public ApplicationEntry(
      final long lowestPosition, final long highestPosition, final ByteBuffer data) {
    this(lowestPosition, highestPosition, new UnsafeBuffer(data));
  }

  @Override
  public BufferWriter toSerializable(final int term, final RaftEntrySerializer serializer) {
    return new SerializableAdapter(term, serializer, this);
  }

  // didn't make it a record as it's not a data holder but really meant to serialize things, though we haven't fully decided as a team how we want to use records :)
  private static final class SerializableAdapter implements BufferWriter {
    private final int term;
    private final RaftEntrySerializer serializer;
    private final ApplicationEntry entry;

    private SerializableAdapter(final int term, final RaftEntrySerializer serializer, final ApplicationEntry entry) {
      this.term = term;
      this.serializer = serializer;
      this.entry = entry;
    }

    @Override
    public int getLength() {
      return serializer.getApplicationEntrySerializedLength(entry);
    }

    @Override
    public void write(final MutableDirectBuffer buffer, final int offset) {
      serializer.writeApplicationEntry(term, entry, buffer, offset);
    }
  }
}
```

File: journal/src/main/java/io/camunda/zeebe/journal/RecordDataWriter.java (hunk 2 above)
Comment (megglos) [reply journal/src/main/java/io/camunda/zeebe/journal/RecordDataWriter.java:12]: nice that fits perfectly!

**Comment by megglos** [comment]
> 👍
> 
> If I understand correctly, the idea is to not pass a buffer, but to pass a serialization strategy instead? Something like a type which knows how to self-serialize?
> 

Exactly, like you proposed in the linked issue #6671 just without generics. From my perspective it's basically replacing the buffer with something that produces the content of the former buffer on demand without requiring to store the buffer content intermediatly.

> I've never been a big fan of this approach (I'm aware we use it heavily in Zeebe 😄), and would rather split serialization strategy from data, but I can see how here it would become somewhat cumbersome (you'd have to pass the instance and it's serializer around all the way down to the final write operation). I think it's worth exploring in a time-boxed fashion, because maybe I'm overestimating the ugliness of the approach...perhaps you've already done that, I'd be curious to hear what came out of it. Anyway, ignoring this as I agree it's probably not ideal with our current design...
> 

I don't think it's too bad tbh, what are the downsides of it from your perspective? Passing down both separately into the journal is worse from my perspective as it exposes the raft entry objects down into the journal layer, while actually there is no need to care about them there. Alternatively one could just pass down the data object and retrieve the serialiser when needed but that adds even more responsibility to the journal layer.

All that matters for the journal is that there is some data block to be written, right? That's why I think just having a type that offers an interface to write data to some target buffer as well as predict the size of it is actually simple enough.

> We could add a method in `RaftEntry` (which all concrete entries implement), e.g. `BufferWriter toSerializable(int,RaftEntrySerializer)`, which could return a simple adapter encapsulating the `RaftEntrySerializer` and the entry itself. This avoids having to iterate of the possible types, and delegates to the type itself to produce the adapter.
> 

I like that idea, will adopt the impl to eliminate the iteration implemented in the Serializer and spread the wiring to the actual entry classes, making the serialiser simpler!

**Comment by npepinpe** [comment]
> I don't think it's too bad tbh, what are the downsides of it from your perspective? Passing down both separately into the journal is worse from my perspective as it exposes the raft entry objects down into the journal layer, while actually there is no need to care about them there. Alternatively one could just pass down the data object and retrieve the serialiser when needed but that adds even more responsibility to the journal layer.

Coupling data (or your model) to serialization introduces logic required by consumers into the actual type. In this instance I think it's mitigated because we pass in adapters (or kind of view-models), but in others (e.g. the protocol) you end up using your classes as both data holders/value classes and as serializers. It's convenient somewhat, but it tends to muddy up the semantics on how to use that class, and can sometimes carry certain risks (e.g. if you have zero copy semantics, and serialization somehow changes your internal buffer position, pr a concurrent access to a getter could mess up the serialization process, etc.). Having them separated makes it more obvious what each class is doing and how it should be used.

That said, that's a lot of talk for an issue which is at best an inconvenience :smile:

**Comment by megglos** [comment]
ok got your point! The advantage of the current implementation is that we actually have the data classes decoupled from serialisation code. 🤔 

To eliminate the iteration of types I could replace it with a simple lookup. Maybe best to extract that to something like a `RaftLogEntryBufferWriterProvider`. This way we keep the entry classes simple, wdyt?

**Comment by npepinpe** [comment]
I think in this case, like I said, it's sort of mitigated if we return an adapter class, so the only logic of the data class is that it knows which serializer to call (but not how to serialize). Whether the adapter lives as an internal class or not is not super important, as it can always be extracted if it grows. I think the issue is more when both are conflated.

**Comment by megglos** [comment]
> I think in this case, like I said, it's sort of mitigated if we return an adapter class, so the only logic of the data class is that it knows which serializer to call (but not how to serialize). Whether the adapter lives as an internal class or not is not super important, as it can always be extracted if it grows. I think the issue is more when both are conflated.

Ok, got ya, moved the logic to the entry classes with [this commit](https://github.com/camunda/zeebe/pull/10320/commits/69a4ec438962e9bf86f7a117d17ecc36a88bfe85)

Thanks for the input!

File: atomix/cluster/src/main/java/io/atomix/raft/storage/serializer/RaftEntrySerializer.java [hunk 3]
```diff
@@ -62,8 +88,30 @@
...
-   * @param buffer
+   * @param buffer to read the raft log entry from
    * @return RaftLogEntry
    */
   RaftLogEntry readRaftLogEntry(DirectBuffer buffer);
+
+  /**
+   * A buffer writer based on serializer logic provided as functions.
+   *
+   * @param getRecordLengthFunction function that determines the length of the serialized data
```
[the remark this answers is not in the document]
Comment (megglos) [reply atomix/cluster/src/main/java/io/atomix/raft/storage/serializer/RaftEntrySerializer.java]: seems like the code scanner does not properly support javadoc on records. However, I think the javadoc of the params is not really adding value here and removed them.

**Review by deepthidevaki — APPROVED** [review]
Looks good :+1:

File: atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/RaftLogEntry.java [hunk 4]
```diff
@@ -43,4 +44,12 @@ public boolean isInitialEntry() {
   public InitialEntry getInitialEntry() {
     return (InitialEntry) entry;
   }
+
+  public Optional<Long> getLowestASqn() {
```
Comment (npepinpe) [inline atomix/cluster/src/main/java/io/atomix/raft/storage/log/entry/RaftLogEntry.java:48]:
```suggestion
  public Optional<Long> getLowestAsqn() {
```

**Comment by megglos** [comment]
bors merge

**Comment by megglos** [comment]
bors merge

**Comment by megglos** [comment]
bors merge

**Pull request merged**